src/framework/components/system.js

```diff
@@ -131,17 +131,37 @@ pc.extend(pc, function () {
          * This can be overridden by derived Component Systems and either called by the derived System or replaced entirely
          * @param {pc.Component} component The component being initialized.
          * @param {Object} data The data block used to initialize the component.
-         * @param {Array} properties The array of property names of the component.
+         * @param {Array} properties The array of property descriptors for the component. A descriptor can be either a plain property name, or an object specifying the name and type.
          */
         initializeComponentData: function (component, data, properties) {
             data = data || {};
 
             // initialize
-            properties.forEach(function(value) {
-                if (data[value] !== undefined) {
-                    component[value] = data[value];
+            properties.forEach(function(descriptor) {
+                var name;
+                var type;
+
+                // If the descriptor is an object, it will have `name` and `type` members
+                if (typeof descriptor === 'object') {
+                    name = descriptor.name;
+                    type = descriptor.type;
+                // Otherwise, the descriptor is just the property name
                 } else {
-                    component[value] = component.data[value];
+                    name = descriptor;
+                }
+
+                var value = data[name];
+
+                if (value !== undefined) {
+                    // If we know the intended type of the value, convert the raw data
+                    // into an instance of the specified type.
+                    if (type !== undefined) {
+                        value = convertValue(value, type);
+                    }
+
+                    component[name] = value;
+                } else {
+                    component[name] = component.data[name];
```

Comment:
This change allows us to specify types in the `_schema` for a component ([like this](https://github.com/playcanvas/engine/pull/1197/files#diff-ee52ec0f40648c086f4cc8cb607189a5R6)), which then means we don't need to hand-write various bits of type casting (e.g. in `initializeComponentData()`).

Reply:
I can pull it out into a separate PR if you prefer.